storage/src/main/java/org/apache/kafka/server/log/internals/LogConfig.java

```diff
@@ -0,0 +1,504 @@
+                + "inter-broker protocol version `" + interBrokerProtocolVersionString + "` is greater or "
+                + "equal than 3.0. This configuration is deprecated and it will be removed in Apache Kafka 4.0.";
+        }
+    }
+
+    public static class RemoteLogConfig {
+
+        public final boolean remoteStorageEnable;
+
+        public final long localRetentionMs;
+        public final long localRetentionBytes;
+
+        private RemoteLogConfig(LogConfig config, long retentionMs, long retentionSize) {
+            this.remoteStorageEnable = config.getBoolean(TopicConfig.REMOTE_LOG_STORAGE_ENABLE_CONFIG);
+
+            long localLogRetentionMs = config.getLong(TopicConfig.LOCAL_LOG_RETENTION_MS_CONFIG);
+
+            // -2 indicates to derive value from retentionMs property.
+            if (localLogRetentionMs == -2)
+                this.localRetentionMs = retentionMs;
+            else {
+                // Added validation here to check the effective value should not be more than RetentionMs.
+                if (localLogRetentionMs == -1 && retentionMs != -1)
+                    throw new ConfigException(TopicConfig.LOCAL_LOG_RETENTION_MS_CONFIG, localLogRetentionMs,
+                        "Value must not be -1 as " + TopicConfig.RETENTION_MS_CONFIG + " value is set as " + retentionMs);
+
+                if (localLogRetentionMs > retentionMs)
+                    throw new ConfigException(TopicConfig.LOCAL_LOG_RETENTION_MS_CONFIG, localLogRetentionMs,
+                        "Value must not be more than property: " + TopicConfig.RETENTION_MS_CONFIG + " value.");
+
+                this.localRetentionMs = localLogRetentionMs;
+            }
+
+            long localLogRetentionBytes = config.getLong(TopicConfig.LOCAL_LOG_RETENTION_BYTES_CONFIG);
+
+            // -2 indicates to derive value from retentionSize property.
+            if (localLogRetentionBytes == -2)
+                this.localRetentionBytes = retentionSize;
+            else {
+                // Added validation here to check the effective value should not be more than RetentionBytes.
+                if (localLogRetentionBytes == -1 && retentionSize != -1)
+                    throw new ConfigException(TopicConfig.LOCAL_LOG_RETENTION_BYTES_CONFIG, localLogRetentionBytes,
+                        "Value must not be -1 as " + TopicConfig.RETENTION_BYTES_CONFIG + " value is set as " + retentionSize);
+
+                if (localLogRetentionBytes > retentionSize)
+                    throw new ConfigException(TopicConfig.LOCAL_LOG_RETENTION_BYTES_CONFIG, localLogRetentionBytes,
+                        "Value must not be more than property: " + TopicConfig.RETENTION_BYTES_CONFIG + " value.");
+
+                this.localRetentionBytes = localLogRetentionBytes;
+            }
+        }
+    }
+
+    // Visible for testing
+    public static class LogConfigDef extends ConfigDef {
```

Comment:
Removed `define` overrides since we now rely on the statically defined `ServerTopicConfigSynonyms` to retrieve the server config default/synonym.